contracts/AddressProviderNG.vy

```diff
@@ -0,0 +1,290 @@
+# @version 0.3.10
+"""
+@title CurveAddressProvider
+@author Curve.Fi
+@license Copyright (c) Curve.Fi, 2020-2023 - all rights reserved
+@notice An entrypoint contract for Curve's various registries
+"""
+
+event NewEntry:
+    id: indexed(uint256)
+    addr: address
+    description: String[64]
+
+event EntryModified:
+    id: indexed(uint256)
+    version: uint256
+
+event CommitNewAdmin:
+    deadline: indexed(uint256)
+    admin: indexed(address)
+
+event NewAdmin:
+    admin: indexed(address)
+
+
+struct AddressInfo:
+    addr: address
+    description: String[256]
+    tags: DynArray[String[64], 20]
+    version: uint256
+    last_modified: uint256
+
+
+admin: public(address)
+transfer_ownership_deadline: public(uint256)
+future_admin: public(address)
+
+num_entries: public(uint256)
+check_tag_exists: public(HashMap[String[64], bool])
+check_id_exists: public(HashMap[uint256, bool])
+ids: public(DynArray[uint256, 1000])
+tags: public(DynArray[String[64], 1000])
+id_tag_mapping: HashMap[bytes32, bool]
+get_id_info: public(HashMap[uint256, AddressInfo])
+
+
+@external
+def __init__(_admin: address):
+    self.admin = _admin
+
+
+# ------------------------------ View Methods --------------------------------
+
+
+@view
+@external
+def get_address(_id: uint256) -> address:
+    """
+    @notice Fetch the address associated with `_id`
+    @dev Returns empty(address) if `_id` has not been defined, or has been unset
+    @param _id Identifier to fetch an address for
+    @return Current address associated to `_id`
+    """
+    return self.get_id_info[_id].addr
+
+
+@view
+@external
+def get_addresses_for_tags(_tag: String[64]) -> DynArray[address, 100]:
+    """
+    @notice Fetch all addresses that are tagged with `_tag`
+    @param _tag tag to fetch an address for
+    @return Array of addresses tagged with `_tag`
+    """
+
+    tagged_ids: DynArray[address, 100] = []
+    for _id in self.ids:
+        if self.id_tag_mapping[keccak256(concat(uint2str(_id), _tag))]:
+            tagged_ids.append(self.get_id_info[_id].addr)
+
+    return tagged_ids
+
+
+# -------------------------- State-Mutable Methods ---------------------------
+
+
+@external
+def add_new_tags(_new_tags: DynArray[String[64], 20]):
+    """
+    @notice Add new tag
+    @dev method allows a max number of entry of 20
+    @param _new_tags An array of types to add to the registry
+           e.g. ['StableSwap', 'CryptoSwap', ...] 
+    """
+    for _new_tag in _new_tags:
+        if not self.check_tag_exists[_new_tag]:
+            self.check_tag_exists[_new_tag] = True
+            self.tags.append(_new_tag)
```

Comment:
Do we want `remove_tags` as well? Typos are common. We might want to clean up deprecated tags at some point too.